src/api/rate_limiter.py

```diff
@@ -27,122 +49,168 @@ class RateLimiter:
         and giving back the result to the calling functions when it is received (async probably)
     """
 
-    def __init__(self):
+    def __init__(self, request_timeout=None, timeout_delay=None, idle_state_sleep=None):
         self.requests = {}
         self.queue = asyncio.Queue()
         # set a max_retry cap
-        if getenv("MAX_API_RETRY") is not None:
-            self._max_retry = int(getenv("MAX_API_RETRY"))
+        if getenv("MAX_API_ATTEMPT") is not None:
+            self._max_attempt = int(getenv("MAX_API_ATTEMPT"))
         else:
-            self._max_retry = DEFAULT_MAX_RETRY
+            self._max_attempt = DEFAULT_MAX_ATTEMPT
 
+        self._pause = False
+        self._idle = False
+        self._request_timeout = request_timeout or DEFAULT_REQUEST_TIMEOUT
+        self._timeout_delay = timeout_delay or DEFAULT_TIMEOUT_DELAY
+        self._idle_state_sleep = idle_state_sleep or DEFAULT_IDLE_STATE_SLEEP
         self.task = asyncio.create_task(self.handle_requests())
-
         self.logger = logging.getLogger(__name__)
 
+    def go_idle(self):
+        self._idle = True
+        self.logger.error("Entering idle state due to previous errors")
+
+    def is_idle(self):
+        return self._idle
+
+    def exit_idle(self):
+        self.logger.info("Resuming from idle state")
+        self._idle = False
+
+    def is_paused(self):
+        return self._pause
+
+    async def pause(self, delay):
+        self._pause = True
+        await asyncio.sleep(delay)
+        self._pause = False
+
+    def handle_get_request(self, request):
+        try:
+            # TODO: we should really use a session object to persist cookies and TCP connection
+            # https://requests.readthedocs.io/en/latest/user/advanced/#session-objects
+            resp = requests.get(request.url, cookies=request.cookies, timeout=self._request_timeout)
+        except requests.exceptions.Timeout as exc:
+            raise RLErrorWithPause(
+                request, self._timeout_delay, self.logger.error, "Timeout"
+            ) from exc
+
+        match resp.status_code:
+            case 200:
+                return resp.json()
+            case 429:
+                try:
+                    time_to_wait = int(resp.headers["Retry-After"])
+                except (KeyError, ValueError) as exc:
+                    raise RateLimiterError(
+                        request,
+                        self.logger.error,
+                        "Too many requests (429) and cannot parse headers",
+                    ) from exc
+
+                raise RLErrorWithPause(
```

Comment:
time_to_wait is define l.103 in the try, so if there is an exception with the header reading, won't it crash when trying to execute the raise RLErrorWithPause with a time_to_wait outdated or uninitialized ?

Reply:
That's why we `except` this case:
if the parsing fail, we raise an exception **without** `time_to_wait` L105 (and it "exits" the fucntion)
if it succeeds we end up here and raise an exxcpetion **with** it